src/Data/Iterable.purs

```diff
@@ -0,0 +1,70 @@
+module Data.Iterable where
+
+import Prelude
+import Control.Monad.Eff (Eff, foreachE)
+import Data.Foldable (traverse_)
+import Data.Traversable (traverse)
+import Unsafe.Coerce (unsafeCoerce)
+
+import Data.Maybe (Maybe)
+import Data.Maybe.First (First)
+import Data.Maybe.Last (Last)
+import Data.Monoid.Additive (Additive)
+import Data.Monoid.Conj (Conj)
+import Data.Monoid.Disj (Disj)
+import Data.Monoid.Dual (Dual)
+import Data.Monoid.Multiplicative (Multiplicative)
+
+-- | `Iterable` represents data structures which can be _traversed_,
+-- | within the Eff monad.
+-- |
+-- | - `forEach` runs an effect for every element in a data structure,
+-- |   and accumulates the results, in the manner of `traverse`.
+-- | - `forEach_` runs an effect for every element in a data structure,
+-- |   without result communication, in the manner of `traverse_`.
+-- |
+-- | While `forEach` and `forEach_` should be functionally equivalent
+-- | to `traverse` and `traverse_`, respectively. The `Iterable` typeclass
+-- | is provided such that optimized versions of these functions may
+-- | be written for the Eff monad. For an example, see `iterableArray`,
+-- | below.
+
+class Iterable i where
```

Comment:
If this approach is still part of the discussion, then I'm thinking something like `FoldableE` and `TraversableE`, since really what is being proposed are just optimized versions for Eff of these two classes, and this follows the precedent of `foreachE`.